snow/engine/snowman/event/queue.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package event
+
+import (
+	"context"
+
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+type Job interface {
+	Execute(context.Context) error
+	Cancel(context.Context) error
+}
+
+type job[T comparable] struct {
+	// If empty, the job is ready to be executed.
+	dependencies set.Set[T]
+	// If true, the job should be cancelled.
+	shouldCancel bool
+	// If nil, the job has already been executed or cancelled.
+	job Job
+}
+
+type Queue[T comparable] struct {
+	// jobs maps a dependency to the jobs that depend on it.
+	jobs map[T][]*job[T]
```

Comment:
I think `dependents` is a clearer name than `jobs` (which I originally assumed to mean `dependencies`).